templates/puppet/integration_spec.rb

```diff
@@ -0,0 +1,130 @@
+- don't need to use any functions that are loaded from modules
+- ensure all our dependent modules have their 'lib' dir in $LOAD_PATH
+- do all our own variable substitution
+- rewrite any necessary functions in puppet and stick them in
+    the manifest.
+- import all the magic from puppetlabs_spec_helper/module_spec_helper,
+    which I have done without fully understanding it.  The comments in
+    the code suggest that the authors also do not fully understand it.
+
+Hopefully that explains the complexity here in a way that will continue
+to be useful as the tests develop further.
+
+Let me also explain the order in which these commands are run.  First,
+there is a pre-destroy - in the event that an older test failed to clean
+up after itself, the next run should take care of it.  Second, we create
+the resources in the manifest.  Third, we run the create again, mocking
+out any requests during 'flush' and making them crash the test - this is
+to check idempotency of the manifests.  Fourth, we delete the resources
+in the manifest, and fifth, we check the idempotency of that delete by
+confirming that no resources make web requests during 'flush'.
+
+VCR recordings are called 'cassettes', and they are stored in
+'spec/cassettes'.  If present, they are used - if absent, created.
+#-%>
+<%= compile 'templates/license.erb' -%>
+
+<%= lines(autogen_notice :puppet) -%>
+require 'spec_helper'
+require 'vcr'
+
+VCR.configure do |c|
+  c.cassette_library_dir = 'spec/cassettes'
+  c.hook_into :webmock
+  c.configure_rspec_metadata!
+end
+
+describe '<%= Google::StringUtils.underscore(obj.name) -%>.create', vcr: true do
+  it 'creates and destroys non-existent <%= Google::StringUtils.underscore(obj.name) -%>' do
+    puts 'pre-destroying <%= Google::StringUtils.underscore(obj.name) -%>'
+    VCR.use_cassette('pre_destroy_<%= Google::StringUtils.underscore(obj.name) -%>') do
+      run_example('delete_<%= Google::StringUtils.underscore(obj.name) -%>')
+    end
+    puts 'creating <%= Google::StringUtils.underscore(obj.name) -%>'
+    VCR.use_cassette('create_<%= Google::StringUtils.underscore(obj.name) -%>') do
+      run_example('<%= Google::StringUtils.underscore(obj.name) -%>')
+    end
+    puts 'checking that <%= Google::StringUtils.underscore(obj.name) -%> is created'
+    VCR.use_cassette('check_<%= Google::StringUtils.underscore(obj.name) -%>') do
+      validate_no_flush_calls('<%= Google::StringUtils.underscore(obj.name) -%>')
+    end
+    puts 'destroying <%= Google::StringUtils.underscore(obj.name) -%>'
+    VCR.use_cassette('destroy_<%= Google::StringUtils.underscore(obj.name) -%>') do
+      run_example('delete_<%= Google::StringUtils.underscore(obj.name) -%>')
+    end
+    puts 'confirming <%= Google::StringUtils.underscore(obj.name) -%> destroyed'
+    VCR.use_cassette('check_destroy_<%= Google::StringUtils.underscore(obj.name) -%>') do
+      validate_no_flush_calls('delete_<%= Google::StringUtils.underscore(obj.name) -%>')
+    end
```

Comment:
On the first part, that sounds fine. I'm a little unclear how we're checking the cassettes to make sure they work properly (are we checking by hand? when will the tests "crash" because the network calls differ? what about headers, access tokens, and the like?), but I'm assuming that'll all become more clear in due time.

On the second part, I've got a lot of thoughts, some of which might be tangential:

I think we can do this using good old-fashioned before/after test calls on RSpec.

In the case of RRS - before we run the tests, we run the create-managed-zone example. After we run all of the tests, we run the delete-managed-zone example. Rspec has [before/after test handlers](https://relishapp.com/rspec/rspec-core/v/2-2/docs/hooks/before-and-after-hooks) that will make the tests look nice + handle running things before/after for us (you've probably seen/used those before).

This is assuming that the create/delete managed-zone examples use the same name as the create/delete resource-record-set.

Now, as for coming up with the dependency patterns, we shouldn't let Puppet figure that out. Magic Modules can pretty easily figure that out though. If you look at the autogenerated Chef tests, you'll see that each autogenerated test is placed the proper order. That's because Chef doesn't build out a graph like Puppet does. If you put the Chef blocks in the wrong order, things just crash. I built out a dependency graph that looks at all of the ResourceRefs and figures out what should be run in what order.

We could use that idea and just force our tests to run in a certain order.

Reply:
Interesting.  That's very handy!  I bet that will work great!  What a fun idea.  I'll poke at it next week - I need tomorrow for some other work.

Reply:
I've thought about this some more, and I'm not inclined to implement it the way you suggested at first glance.  Let me see if we're on the same page about this.

Problem with the current implementation:
* dangling resources are possible, or even likely, or even literally-impossible-to-avoid in some resources.
* every test also implicitly tests its dependencies - failures won't be isolated, making root-causing challenging.

But I see some issues with your methodology as well:
* all examples are tied together, meaning that a change to improve testing of `network` will require 10-50 changes throughout the examples directory.
* all examples need to use the same dependencies, which means that we'll have a real hard time with, e.g. VPN Tunnel testing.
* it's not possible to test autorequires - we're forcing create-before-create and destroy-after-destroy behavior that we actually want to confirm works correctly.
* lamely: dependency graph behavior feels complicated and I'm inclined to not add complicated things, because debugging tests is already a pain and I don't want to make it more confusing in there.

So here's my theory.  I could rewrite the entire set of examples, splitting them into dependencies and unit-under-test - this is almost done already, honestly, with the 'filename != readme' check that's in most of 'em.  Then we could know which tests failed in the `before` bit and which tests failed in the unit-under-test bit.  Then we get rid of the dependency on a `delete_foo` manifest and just swap the present to absent bit in the unit-under-test, then swap the present to absent bits in the dependencies.  4-step process, most of the perks, few of the flaws.

Downside: that sounds hard to do.
Question: is it worth doing?

Reply:
Totally understand the issues.

I don't entirely follow your solution